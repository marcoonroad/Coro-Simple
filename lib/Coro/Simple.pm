use v6;

unit module Coro::Simple;

sub coro (&block) is export {
    return sub (*@params) {
        # we explore the lazy evaluation property
        # to evaluate the computation on demand
        my @yields := (gather { block |@params }).list;
        my $index   = 0;
        # generator
        return sub ( ) {
            my $result;
            if defined @yields[ $index++ ] {
                $result = @yields[ $index - 1 ];
            }
            # otherwise, if the coroutine is dead
            else {
        	$index -= 1; # disallows further lookups
                $result = False;
            }
            return $result;
        }
    }
}

sub yield ($value) is export { take $value }

sub suspend( ) is export { take True }

# purpose:
# to check if generated value was returned from yield triggering (rather end of computation)
sub ensure (&block, $value) is export {
    if ($value ~~ Bool) && (!$value) {
        return block;
    }
    return $value;
}

# deprecated function
sub assert (&block, $value) is export {
    warn "[ Deprecated function ]\n" ~
        "This function will be removed in a future version. Please use 'ensure' instead.";
    return ensure &block, $value;
}

sub from (&generator) is export {
    my $temp = generator; # obtains the first value

    return gather {
        while ($temp !~~ Bool) || (?$temp) {
            take $temp;
            $temp = generator;
        }
    };
}

# end of module
