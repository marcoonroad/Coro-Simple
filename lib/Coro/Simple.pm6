use v6;

unit module Coro::Simple;

sub coro (&block) is export {
  my $coroutine-thread = $*THREAD;

  return sub (*@params) {
    # we explore the lazy evaluation property
    # to evaluate the computation on demand
    my @yields := (gather { block |@params }).list;
    my $index   = 0;
    # generator
    return sub ( ) {
      unless $*THREAD.id == $coroutine-thread.id {
        die "Yield called outside the owner coroutine!";
      }

      my $result;
      if defined @yields[ $index++ ] {
        $result = @yields[ $index - 1 ];
      }
      # otherwise, if the coroutine is dead
      else {
	      $index -= 1; # disallows further lookups
        $result = False;
      }
      return $result;
    }
  }
}

sub yield ($value) is export { take $value }

sub suspend( ) is export { take True }

# purpose:
# to check if generated value was returned from yield triggering (rather end of computation)
sub ensure (&block, $value) is export {
  if ($value ~~ Bool) && (!$value) {
    return block;
  }
  return $value;
}

sub from (&generator) is export {
  my $temp = generator; # obtains the first value

  return gather {
    while ($temp !~~ Bool) || (?$temp) {
      take $temp;
      $temp = generator;
    }
  };
}

# end of module
