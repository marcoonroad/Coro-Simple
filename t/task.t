#!/usr/bin/perl6

use v6;

# a task test

use Test;
use Coro::Simple;

plan 2;

my $expected = slurp './t/expected/task.out';
my $handler = open './t/expected/task.out', :w;

my @tasks;

sub add-task ($block) {
  my $coro = coro $block;
  @tasks.push: $coro( );
}

sub dispatch-now ( ) {
  while ?@tasks {
    my $task = @tasks.shift;
    @tasks.push: $task if $task( );
  }
  True;
}

# add a task and run getting started with it
sub spawn-task ($block) {
  my $coro = coro $block;
  my $task = $coro( );
  @tasks.push: $task if $task( );
  ok dispatch-now, "dispatcher ran successfully";
}

# tasks
add-task {
  for ^3 -> $i {
    $handler.say: "Ping -> $i";
    suspend;
  }
}

add-task {
  my @entries = @("\n", "WTF?", "\n\n");

  for ^5 {
    $handler.say: ([~] @entries);
    suspend;
  }
}

spawn-task {
  for ^8 -> $i {
    $handler.say: "Pong -> $i";
    suspend;
  }
}

$handler.close;
my $current = slurp './t/expected/task.out';
is $expected, $current, "generated logs must match";

# end of test
