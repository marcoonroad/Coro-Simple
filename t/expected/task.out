Pong -> 0
Ping -> 0

WTF?


Pong -> 1
Ping -> 1

WTF?


Pong -> 2
Ping -> 2

WTF?


Pong -> 3

WTF?


Pong -> 4

WTF?


Pong -> 5
Pong -> 6
Pong -> 7
